ntoskrnl/mm/ARM3/pagfault.c

```diff
@@ -1919,19 +1917,22 @@ _WARN("Session space stuff is not implemented yet!")
             /* Use the session process and working set */
             CurrentProcess = HYDRA_PROCESS;
             WorkingSet = &MmSessionSpace->GlobalVirtualAddress->Vm;
-
-            /* Make sure we don't have a recursive working set lock */
-            if ((CurrentThread->OwnsSessionWorkingSetExclusive) ||
-                (CurrentThread->OwnsSessionWorkingSetShared))
-            {
-                /* Fail */
-                return STATUS_IN_PAGE_ERROR | 0x10000000;
-            }
         }
 
-        /* Acquire the working set lock */
-        KeRaiseIrql(APC_LEVEL, &LockIrql);
-        MiLockWorkingSet(CurrentThread, WorkingSet);
+        /* Acquire the working set lock.
+         * Callers forcing page-in operation (not a real page fault then)
+         * must already hold it*/
+        if (TrapInformation)
+        {
+
+            KeRaiseIrql(APC_LEVEL, &LockIrql);
+            MiLockWorkingSet(CurrentThread, WorkingSet);
+        }
+        else
+        {
+            /* IRQL was already checked above */
+            ASSERT(IsSessionAddress ? PsGetCurrentThread()->OwnsSessionWorkingSetExclusive : PsGetCurrentThread()->OwnsSystemWorkingSetExclusive);
+        }
```

Comment:
Two things:
1. I wouldn't say that passing bogus pointers to a function is fine.
2. This avoids unlocking and locking again each time we call MmAccessFault. And when you do such things, you're supposed to read again any paging structure that you read before, because the stuff you just read might have disappear under the action of the modified page writer, the working set trimmer or anything else freeing or allocating virtual Memory. So "proper locking & unlocking" is full of traps in this case.

The hack added in `MiMakeSystemAddressValid` is removed in a later commit.

Reply:
Let's see. These are the places, where MmAccessFault is called from:

- Page fault handler: unlocked
- MmProbeAndLockPages (twice): unlocked
- MiMakeSystemAddressValid: unlocks and relocks
- MiMakeSystemAddressValidPfn: unlocks and relocks
- MiLockVirtualMemory: Aquires lock, releases it for MmAccessFault

Here's what MmAccessFault looks like:
```
MmAccessFautlt(...)
{
    if (Address >= MmSystemRangeStart)
    {
        MiSynchronizeSystemPde()
        MiCheckPdeForPagedPool()

        // Shouldn't be done for user PDEs!
        if (PointerPde->u.Hard.Valid == 0)
        {
            KeBugCheck();
        }


        TempPte = *PointerPte;
        if (TempPte.u.Hard.Valid == 1)
        {
            // not relevant
        }

        if (MI_IS_SESSION_PTE(Address))
        {
            MiCheckPdeForSessionSpace
        }

        if (MI_IS_PAGE_TABLE_OR_HYPER_ADDRESS(Address))
        {
            goto UserFault;
        }

        //Get correct workingset/process (session/normal)
        MiLockWorkingSet(CurrentThread, WorkingSet);

        // All the things locked kernel mode

        MiUnlockWorkingSet(CurrentThread, WorkingSet);
        return Status;
     }

UserFault:

    MiLockProcessWorkingSet(CurrentProcess, CurrentThread);

    // All the things locked user mode

    MiUnlockProcessWorkingSet(CurrentProcess, CurrentThread);
    return Status
}
```
There is actually a bug: We call MiMakeSystemAddressValid to page in user mode PDEs etc. This goes through the kernel mode path in MmAccessFault, which then bugchecks, when higher level PDEs are not present. This is wrong.

Suggested code change:

1. Factor out "All the things locked" for both kernel and user
2. Call the kernel thing from MiMakeSystemAddressValid (probably needs the extra 
3. Introduce MiMakePageTableAddressValid and call the user path from that
4. Call the user thing from MiLockVirtualMemory

This would make the page fault handler code much more readable, fix the paging in of PDEs, and remove unnecessary code pathes.

Reply:
>     * Page fault handler: unlocked
OK
>     * MmProbeAndLockPages (twice): unlocked
This is wrong. The WS lock must be held all along, otherwise the pages at the beginning of the probed range can be paged out by the time we're done with probing and all hell will break loose when actually locking the pages.
>     * MiMakeSystemAddressValid: unlocks and relocks
And warns caller about lock state change.
>     * MiMakeSystemAddressValidPfn: unlocks and relocks
Ditto
>     * MiLockVirtualMemory: Aquires lock, releases it for MmAccessFault
Changed in this PR.

Reply:
Ok, but the refactoring I suggested can be used for those functions as well.

Reply:
working on it. This is not trivial :slightly_smiling_face: